File: java/src/leetcode/p240/Solution.java
package leetcode.p240;

class Solution {
    public static void main(String[] args){
        int[][] arg = {{1,3,5}};//{{1,4,7,11,15},{2,5,8,12,19},{3,6,9,16,22},{10,13,14,17,24},{18,21,23,26,30}};
        Solution sol = new Solution();
        boolean res = sol.searchMatrix(arg,-1);
        System.out.println(res);
    }

    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null) return false;
        if (matrix.length == 0) return false;
        if (matrix[0].length == 0) return false;
        int n = matrix.length;
        int m = matrix[0].length; //not checking that its truely square n*m
        //return recurse_search(0,n*m - 1,target,matrix,n,m);

        return threequartier(matrix,0,0,n-1,m-1,target);
    }


    boolean threequartier(int [][]matrix, int sx, int sy, int ex, int ey,int target){
        if (sx > ex || sy > ey || sx < 0 || sy < 0 || ex >= matrix.length || ey >= matrix[0].length) return false;

        int xsize = ex - sx;
        int ysize = ey - sy;

        if (xsize == 0 && ysize == 0)
            if (matrix[sx][sy] == target) return true;
            else return false;
        if (xsize == 1 && ysize == 1){
            if (matrix[sx][sy] == target || matrix[sx+1][sy] == target || matrix[sx][sy+1] == target || matrix[sx+1][sy+1] == target) return true;
            else return false;
        }
        if (xsize == 1 && ysize == 0)
            if (matrix[sx][sy] == target || matrix[sx+1][sy] == target) return true;
            else return false;
        if (xsize == 0 && ysize == 1)
            if (matrix[sx][sy] == target || matrix[sx][sy+1] == target) return true;
            else return false;

        int midx = sx + (ex - sx)/2;
        int midy = sy + (ey - sy)/2;

        int cur = matrix[midx][midy];
        if (cur == target)return true;

        if (cur > target){
            return (threequartier(matrix,sx,sy,midx,midy,target) || threequartier(matrix,midx + 1,sy,ex,midy-1,target)  ||
                    threequartier(matrix,sx,midy +1,midx-1,ey,target) );
        } else {
            return (threequartier(matrix,midx,midy,ex,ey,target) || threequartier(matrix,midx + 1,sy,ex,midy-1,target)  ||
                    threequartier(matrix,sx,midy +1,midx-1,ey,target) );
        }
    }




}

File: scala/src/leetcode/p240/Solution.scala
package leetcode.p240

import scala.annotation.tailrec


object Solution {
    def searchMatrix(matrix: Array[Array[Int]], target: Int): Boolean = {
        binsearch(0,matrix.size, 0, matrix(0).size, matrix, target) 
    }
    
    def binsearch(l: Int, r: Int, u: Int, d: Int, matrix: Array[Array[Int]], target: Int): Boolean = {
        if(r <= l || d <= u){
            return false
        } else{
            val idx = l + (r - l - 1)/2
            val idy = u + (d - u - 1)/2

            val check = matrix(idx)(idy)

            if (check == target){
                return true
            } else {
                if (check < target){
                    binsearch(idx + 1 , r, idy + 1, d, matrix, target) ||
                    binsearch(l, idx + 1, idy + 1, d, matrix, target) ||
                    binsearch(idx + 1, r, u, idy + 1, matrix, target )
                } else{
                    binsearch(l, idx, u, idy, matrix, target) ||
                    binsearch(idx, r, u, idy, matrix, target) ||
                    binsearch(l, idx, idy, d, matrix, target) 
                }
            }
        }
    }
}


object test{
    @main
    def main() = {
        val m = Array(Array(1), Array(3), Array(5))
        //val m = Array(Array(1,4,7,11,15),Array(2,5,8,12,19),Array(3,6,9,16,22),Array(10,13,14,17,24),Array(18,21,23,26,30))
        println(Solution.searchMatrix(m, 1))
    }

}

